# lncRNA
## 1.Build hisat2 index
```bash
mamba activate hisat2.2.1
cd 11.lncRNA
nohup hisat2-build Ds-v1.fa 02.index/Ds-v1_hisat2.2.1 &
```
## 2.mapping to the genome
```bash
#hisat2 version 2.2.1
cd 11.lncRNA
cat sample_list | while read i
do
  nohup hisat2 -x 02.index/Ds-v1_hisat2.2.1 --dta -p 10 --rna-strandness RF -1 01.CleanData/"$i".R1.fq.gz -2 01.CleanData/"$i".R2.fq.gz 2> 03.bam/"$i".hisat2.log > 03.bam/"$i".sam &
done
#samtools version 1.19
cat sample_list | while read i
do
nohup samtools sort -@ 10 03.bam/$i.sam -o 03.bam/$i.sort.bam &
done
for i in `cat sample_list`;do nohup samtools view -@ 10 -b -q 50 03.bam/$i.sort.bam >03.bam/$i.sort.uq.bam &;done
```
## 3.obtain new annotation sites
```bash
#stringtie version 2.2.1
cd 11.lncRNA
for i in `cat sample_list`;do nohup stringtie 03.bam/$i.sort.uq.bam -o 04.GTF/$i.gtf -p 20 -G /storage/lihongxiang/Ds-v1_project/0.Ds-v1_Genome/Ds-v1.format.gff3 -l $i &;done
```
## 4.merge all the gtf files by Cufflinks
```bash
mamba activate cufflinks
#cuffmerge --version
#merge_cuff_asms v1.0.0
#cufflinks v2.2.1
cuffmerge -p 30 -g Ds-v1.format.gtf -s Ds-v1.fa -o merge_gtf gtf_list
```
## 5.screen new transcript loci with class code "u" and the number of exon exceeds 1
```bash
grep ' class_code "u"' merged.gtf >merged_u.gtf
grep 'exon_number "2"' merged_u.gtf >merged_u_exon_2.gtf
cut -f 9 merged_u_exon_2.gtf|cut -f 2 -d ';' >transcript_id
grep -f transcript_id merged_u.gtf >merged_u_exon_more_than_2.gtf
```
## 6.transfer gtf file to fa file
```bash
gffread -w 05.fa/merged_select.fa -g ~/Project/0.Ds-v1_Genome/Ds-v1.fa 04.GTF/merge_gtf/merged_u_exon_more_than_2.gtf
```
## 7.identify lncRNAs
use lncfinder2.0ï¼Œhttps://github.com/hanlinqian/LncRNA_Finder2.0?tab=readme-ov-file
```
#The script first screens by transcript length and ORF, and obtains lnc_DAF145.newlnc-RNAlen_ORFlen_info.txt and lnc_DAF145.newlnc-lenfilterleft.fa
#Then use CPC2 (CPC2: a fast and accurate coding potential calculator based on sequence intrinsic features) (http://cpc2.cbi.pku.edu.cn) program to screen, and obtain lnc_DAF145.newlnc-CPC_left.fa and lnc_DAF145.newlnc-cpc_res_table.txt.txt
#Finally, use housekeeping sequences, that is, other ncRNA sequences, including rRNA, tRNA, snRNAs and snoRNAs, etc., to screen by blast, and the comparison result is lnc_DAF145.newlnc-housekeeping_blast_res.txt, and delete the sequences with high comparison rate with housekeeping sequences
perl LncRNA_Finder2.0.pl -i 11.lncRNA/05.fa/merged_select.fa -k housekeeping.fa -o 11.lncRNA/06.LncRNA_Finder2.0/lnc.newlnc
#The final result is lnc_DAF145.newlnc-putativelncRNA.fa
```
## 8.calculate expression of lncRNAs
```
grep '>' lnc.newlnc-putativelncRNA.fa|sed 's/>//g'>../07.lncRNA/lncRNA.id
grep -f lncRNA.id ../04.GTF/merge_gtf/merged.gtf >newlnc.gtf
python format_lnc_gtf2gff3.py newlnc.gtf newlnc.gff3 Pbr
for i in `cat sample_list`;do nohup stringtie 03.bam/$i.sort.uq.bam -p 10 -G 07.lncRNA/newlnc.gff3 -e -A 08.exp/${i}_lnc_exp.tab &;done
```
## 9.merge all the tissues' expression to one file
```
python Merge_TPM.py file_list linc.TPM
```
# TotalRNA
## 1.Quantification
```
#stringtie version 2.2.1
mamba activate lncRNA
for i in `cat sample_list`;do nohup stringtie 02.bam/$i.sort.uq.bam -p 10 -G Ds-v1.gtf -e -A 03.exp/${i}_total_exp.tab &;done
python Merge_TPM.py file_list TotalRNA.TPM
```


