# MicroRNA
Identification of microRNA using miRDeep-P2
Article: https://doi.org/10.1093/bioinformatics/bty972
Process: https://github.com/TF-Chan-Lab/miRDeep-P2_pipeline
Software download address: https://sourceforge.net/projects/mirdp2/files/
Process recommended version 1.1.4
## 1.index
```
cd 13.sRNA/miRDP2/1.index
#构建索引
nohup bowtie-build Ds-v1.fa ./Ds-v1-bowtie1.2.2 &
```
## 2.run miRDP2-v1.1.4_pipeline.bash
```
cd 13.sRNA/miRDP2
bowtie-build -f ./ncRNA_rfam.fa 1.1.4/scripts/index/rfam_index
#Process reads, remove duplicates, and change the head to the required format
cd 13.sRNA/04.uniq_seq
for i in `cat sample_list`;do nohup python srnauniq_for_midp2.py ../03.trim18-26/sRNA_${i}.18-26.fa $i >$i.midp2.fa &;done
#运行miRDP2-v1.1.4_pipeline.bash
for i in `cat sample_list`;do nohup bash ./1.1.4/miRDP2-v1.1.4_pipeline.bash -g Ds-v1.fa -x ./1.index/Ds-v1-bowtie1.2.2 -f -i 13.sRNA/04.uniq_seq/$i.midp2.fa -o ./2.result &;done
```
## 3.Merging multiple sample results
```
for i in cat ../sample_list;do cat ../2.result/$i.midp2/$i.midp2_filter_P_prediction >>miRNA.midp2_filter_P_prediction;done
sort -k1,1V -k5,5n miRNA.midp2_filter_P_prediction >miRNA.midp2_filter_P_prediction_sort
```
## 4.Making GFF and FA
```
python merge_miRNA_to_gff.py miRNA.midp2_filter_P_prediction_sort Pbr miRNA.gff3 miRNA.fa
```
## 5.Download publicly known mature fa
```
mkdir 13.sRNA/miRDP2/4.mature_fa
cd 4.mature_fa
#Download mature.fa from https://mirbase.org/download/
wget https://mirbase.org/download/mature.fa
#To convert all "U" nucleotide to "T"
perl ../miRDeep-P2_pipeline-main/script/fasta_U2T.pl mature.fa mature_wo_U.fa
#To collapse unique mature sequences in mature_wo_U.fa
perl ../miRDeep-P2_pipeline-main/script/unique_fasta_v1.3.pl mature_wo_U.fa mature_wo_U_uniq.fa mature_uniq
#Index file of the known miRNA sequences, e.g. the miRBase sequences, generated by blast using the following command
makeblastdb -in mature_wo_U_uniq.fa -dbtype 'nucl'
```
## 6.Blast to identify the known parts of the identified miRNA
```
mkdir 13.sRNA/miRDP2/5.blast
cd 5.blast
blastn -db ../4.mature_fa/mature_wo_U_uniq.fa -query ../3.miRDP2_merge/miRNA.fa -out miRNA_blastn.txt -word_size 4 -num_alignments 1
perl ../miRDeep-P2_pipeline-main/script/general_blast_parser.pl miRNA_blastn.txt miRNA_blastn_parsed.txt
perl ../miRDeep-P2_pipeline-main/script/parse_parsed_blast_known_plants.pl miRNA_blastn_parsed.txt miRNA
#获取known miRNA id：
cut -f2 miRNA_known.txt | sort | uniq > miRNA_known_id.txt
#Get variant miRNA id:
cut -f2 miRNA_variant.txt | sort | uniq > temp.txt
#This script will check whether the ID in temp.txt is in miRDP2_mature_known_id.txt, and output all the IDs in temp.txt that are not in miRDP2_mature_known_id.txt. This means that if the variant miRNA ID has already appeared in the known miRNA ID, it will not appear again.
#My result, the two have no intersection, so temp.txt is the same as miRDP2_mature_variant_id.txt
perl ../miRDeep-P2_pipeline-main/script/filter_lines_by_key_words_list.pl temp.txt miRNA_variant_id.txt miRNA_known_id.txt 0
rm temp.txt
#add labels
grep  '>' ../3.miRDP2_merge/miRNA.fa|sed 's/>//g' >all_miRNA_id
```
## 7.count
```
#index
mkdir 6.count
cd 13.sRNA/miRDP2/6.count
mkdir index
cd index
cp ../../3.miRDP2_mature/miRNA.fa .
bowtie-build miRNA.fa miRNA.fa
#mapping
ls 13.sRNA/03.trim18-26/sRNA_DAF*fq >fastq_list.txt
for i in `cat fastq_list.txt`;do bowtie -v 0 --norc -S index/miRNA.fa $i | samtools view -Sb - > ${$(basename $i)%.fq}.miRNA.bam;done
#count
for i in `cat ../sample_list`;do nohup perl ../miRDeep-P2_pipeline-main/script/bam2ref_counts.pl -bam sRNA_${i}.18-26.miRNA.bam -f index/miRNA.fa >${i}_count.txt &;done
#for i in *count.txt;do ls $i >>count_list.txt;done
perl ../../miRDeep-P2_pipeline-main/script/combine_htseq_counts.pl count_list.txt count_table.txt
awk 'NR > 1 && ($2+$3+$4+$5+$6+$7+$8+$9+$10+$11+$12+$13) == 0' count_table.txt|wc -l
#make final results
python make_final_gff3.py ../5.blast/all_miRNA.annotation ../3.miRDP2_merge/miRNA.pre.gff3 miRNA.gff3 miRNA.annotation
sort -k1,1V ../6.count/count_table.txt > miRNA_count.txt
#get TPM
python cal_miRNA_TPM.py miRNA_count.txt miRNA.CPM
```

