# CircRNA
## 1.mapping to the genome
```
bwa index -a bwtsw Ds-v1.fa
#paired-end reads
#bwa mem â€“T 19 ref.fa read1.fq read2.fq > aln-pe.sam
for i in `cat sample_list`;do nohup bwa mem -T 19 -t 10 Ds-v1.fa 12.circRNA/02.fastuniq/$i.uq.R1.fq.gz 12.circRNA/02.fastuniq/$i.uq.R2.fq.gz 1>$i.aln-pe.sam 2>$i.aln-pe.log &;done
```
## 2.identification
```
for i in cat sample_list;do nohup perl CIRI_v2.0.6/CIRI2.pl -I ./03.bwa/$i.aln-pe.sam -O 04.CIRI_out/gtf_complementary/$i.CIRI_out -F Ds-v1.fa -A Ds-v1.format.gtf &;done
```
## 3.quant
```
#bwa index
nohup bwa index -a bwtsw Ds-v1.fa &
#hisat2 index
nohup hisat2-build Ds-v1.fa Ds-v1 &
#prepare config file
vi config
reference:
  fasta: /storage2/lihongxiang/Project/1.Network-v1/12.circRNA/CIRI/05.CIRIquant/Ds-v1.fa
  gtf: /storage2/lihongxiang/Project/1.Network-v1/12.circRNA/CIRI/05.CIRIquant/Ds-v1.format.gtf
  bwa_index: /storage2/lihongxiang/Project/1.Network-v1/12.circRNA/CIRI/05.CIRIquant/bwa_index/Ds-v1.fa
  hisat_index: /storage2/lihongxiang/Project/1.Network-v1/12.circRNA/CIRI/05.CIRIquant/hisat_index/Ds-v1
#run
for i in `cat sample_list`;do nohup CIRIquant -t 10 -1 12.circRNA/02.fastuniq/$i.uq.R1.fq.gz -2 12.circRNA/02.fastuniq/$i.uq.R2.fq.gz --config ./config -o result/$i -p $i &;done
```
## 4.merge
```
#merge bed file
cd 12.circRNA/CIRI/05.CIRIquant/result
for i in `cat sample_list`;do cat $i/$i.bed >>../../06.merge/circRNA.bed;done
#Merge circRNAs detected in multiple samples and remove duplicates
sort circRNA.bed|uniq > circRNA.merged.uniq.bed
#Input the merged bed and do the quantification again
for i in `cat sample_list`;do nohup CIRIquant -t 30 -1 12.circRNA/02.fastuniq/$i.uq.R1.fq.gz -2 12.circRNA/02.fastuniq/$i.uq.R2.fq.gz --config 12.circRNA/CIRI/05.CIRIquant/config -o $i -p $i --bed circRNA.merged.uniq.bed &;done
#Get the new result, put all gtf files into the gtf folder, and extract CPM
python get_attributes_for_gff3.py circRNA.merged.uniq.bed gtf_list circRNA.gff3
python make_circ_CPM.py circRNA.merged.uniq.bed gtf_list circRNA.CPM
awk -v OFS='\t' 'NR==1{print "ID\t" $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18} NR>1{print "Pbr_circRNA" NR-1, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18}' circRNA.CPM >t && mv t circRNA.CPM
```
