# 3D Genome Network
## 1.Processing Hi-C data using HiC-Pro
```
#HiC-Pro 2.11.1
nohup HiC-Pro --input data --output hic-pro_results --conf config-hicpro.txt &
```
## 2.Using fithic2 to detect loops
```
#Fit-Hi-C 2.0.8
nohup fithic -i fithic.interactionCounts.gz -f fithic.fragmentMappability.gz -t fithic.biases.gz -o ./fithic_intraOnly -r 5000 -L 4000 -U 2000000 -v -x intraOnly &
nohup fithic -i fithic.interactionCounts.gz -f fithic.fragmentMappability.gz -t fithic.biases.gz -o ./fithic_interOnly -r 5000 -v -x interOnly &
```
## 3.Extracting the connections between genes
```
nohup python Hi-C_contacts_network.py -G Pbr.gff -L ./FitHiC.spline_pass1.res5000.significances_contact-8_q-0.01.txt -r 5000 -o Hic_intra_network.csv &   
nohup python Hi-C_contacts_network.py -G Pbr.gff -L ./FitHiC.spline_pass1.res5000.significances_contact-10_q-0.01.txt -r 5000 -o Hic_inter_network.csv &
-L, --loop, fit-hic2 result file: FitHiC.spline_pass1.res5000.significances_contact-10_q-0.01.txt
-G, --gff, gff file: gff file extracted by prepare_gff-v2.pl script
-r, resolution, example: 5000
-o, output file name, csv format
```
## 4.Merge two networks
```
python merge_two_network.py -n1 Hic_inter_network.csv -n2 Hic_intra_network.csv -o Hic_network.csv
```
