# ATAC-PWM Network
## 1.index
```
nohup bowtie2-build --threads 30 DS-v1.fasta DS-v1 &
```
## 2.mapping
```
nohup bowtie2 -x DS-v1 -X 1000 -I 10 -p 30 -q --dovetail --no-unal --very-sensitive-local --no-mixed --no-discordant -1 DS-1_1P.fq.gz -2 DS-1_2P.fq.gz 2> DS-1.alignment.summary|samtools view -@ 30 -F 4 -u - |samtools sort -@ 30 -o DS-1.bam - &
```
## 3.Filter low-quality alignments and remove PCR duplicates
```
samtools view -b -f 2 -q 30 -o ${sample}.f2.q30.bam ${sample}.bam
nohup sambamba markdup -r -p -t 30 DS-2.f2.q30.bam DS-2.f2.q30.dedup.bam 2>DS-2_rmdup.log &
nohup bedtools bamtobed -i DS-1.f2.q30.dedup.bam  > DS-1.final.bed &
```
## 4.callpeak
```
nohup macs2 callpeak -t DS-2.final.bed -n DS-2 --shift -75 --extsize 150 --nomodel -B  -g 512181415 --keep-dup all -q 0.01 --outdir ./DS-2_macs2_results &
```
## 5.Merge two duplicates using idr
```
nohup idr --samples ./Pbr-1_macs2_q_0.01_results/Pbr-1_peaks.narrowPeak ./Pbr-2_macs2_q_0.01_results/Pbr-2_peaks.narrowPeak --peak-list ./Pbr-4.7_macs2_q_0.01_results/Pbr-4.7_peaks.narrowPeak --input-file-type narrowPeak --output-file Pbr-4.7_narrowpeak_q_0.01.idr --rank p.value --soft-idr-threshold 0.05 --plot --use-best-multisummit-IDR &
sort +0 -1 +1n -2 Pbr-4.7.idr.0.05.narrowPeak >Pbr-4.7.idr.0.05.narrowPeak.sorted
awk 'BEGIN{OFS="\t"}{$4="Peak_"NR ; print $0}' Pbr-4.7.idr.0.05.narrowPeak.sorted
```
## 5.Structural annotations of peaks
### Preparation: Import the package, install it if you don't have it.  
In addition, you need to use gff/gtf to build a TxDb. Because the official website does not have TxDb for every species (currently 43 in total), we build it manually.
```R
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("clusterProfiler")

library(clusterProfiler)
library(ChIPseeker)
library(GenomicFeatures)
txdb <- makeTxDbFromGFF("DS.gff",
                        format="gff")
keytypes(txdb)
keys(txdb)
```
### Structural annotations for a single file
```R
#Read a single summits file
peaks <- readPeakFile("Pbr-4.7.idr.0.05.narrowPeak.sorted")
#Check the distribution of all peaks on each chromosome
covplot(peaks, weightCol="FE")
#Structural annotations
#options(ChIPseeker.downstreamDistance = 500)
#Set the length of the gene downstream
peakAnno <- annotatePeak(peaks,
                         TxDb=txdb,
                         level = "gene",
                         tssRegion=c(-2500, 2500))
#After annotation, visualize it and choose from a variety of pictures
plotAnnoBar(peakAnno)
plotDistToTSS(peakAnno)
vennpie(peakAnno)
plotAnnoPie(peakAnno)
#install.packages("ggupset")
library(ggupset)
upsetplot(peakAnno)
#install.packages("ggimage")
library(ggimage)
upsetplot(peakAnno, vennpie=TRUE)
#Finally, convert our annotation results into a data frame for easy viewing
df <- as.data.frame(peakAnno)
write.csv(df,file = '7.6_peakAnno.csv', quote = F, row.names = F)
#Extract the annotated genes (column 14) for subsequent functional analysis
gene <- df[,14]
```
### Structural annotations for multiple files
```R
#You can also read multiple summits files at a time, store them in list, and then annotate them using lapply
files = list(peak1 = "Pbr-4.7.idr.0.05.narrowPeak.sorted", peak2 = ("Pbr-7.6.idr.0.05.narrowPeak.sorted"))
peakAnnoList <- lapply(files, 
                       annotatePeak,
                       TxDb=txdb,
                       level = "gene",
                       tssRegion=c(-2500, 2500))
plotAnnoBar(peakAnnoList)
plotDistToTSS(peakAnnoList)
```
## 6.promoter peaks
### Extract peak sequences from gene promoter regions
```bash
grep Promoter Ds-15_peakAnno.csv > Ds-15_promoter_peak
cut -f 1,2,3,6 -d ',' Ds-15_promoter_peak|awk 'BEGIN{FS=","}{OFS="\t"}{print $1,$2,$3,$4}' >Ds-15_peak.bed
bedtools getfasta -fi Pbr.fasta -bed Ds-15_peak.bed -fo Ds-15_peak.fa -name 
perl -pe 's/(^>\S+?)::.*/$1/g' Ds-15_peak.fa > t && mv t Ds-15_peak.fa
cut -f 6,9,19,20 -d ',' Ds-15_promoter_peak > Ds-15_peak_gene_id
sed -i 's/MERGED:.*; //g' Ds-15_peak_gene_id
```
### TF combined with fimo search
```bash
fimo --o Ds-15_peak Pbr_TF_binding_motifs.meme Ds-15_peak.fa
```
### get promoter peak network
```
python tran_fimo_to_network\(ATAC\).py -f Ds-15_peak/fimo.tsv -p Ds-15_promoter_peak -o Ds-15_promoter_peak.network
```
### Weighting the score
```python
import pandas as pd
import numpy as np 
a = pd.read_csv('Ds-15_promoter_peak.network')
a 
b = a['weight'] 
a['weight'] = (np.log2(b)+1)/(2*np.log2(max(b))+1) + 0.5 a.to_csv('Ds-15_promoter_peak.network',index=False)              
```
## 7.Peak on Hi-C loop
### Extract dOCR peak
```
#In the results of the structural annotation, the peak of the Distal Intergenic region is dOCR, the distal chromatin open region. My screening criteria are 2.5kb upstream of the promoter and 500bp downstream.
grep Intergenic Ds-15_peakAnno.csv|cut -f 1,2,3,6,9 -d ','|awk 'BEGIN{FS=","}{OFS="\t"}{print $1,$2,$3,$4,$5}' >Ds-15_dOCR.peak           
```
### Extract fasta of dOCR on loop
```
python extract_peak_gene_contacts_from_dOCR-gene_loop.py -d dOCR-gene_loop -o Ds-15_peak_gene_contact.id
cut -f 1 Ds-15_peak_gene_contact.id|sort|uniq > Ds-15_loop_dOCR.peak 
grep -wf Ds-15_loop_dOCR.peak Ds-15_dOCR.peak >t && t Ds-15_dOCR.peak
bedtools getfasta -fi Pbr.fasta -bed Ds-15_dOCR.peak -fo Ds-15_dOCR.fa -name
perl -pe 's/(^>\S+?)::.*/$1/g' Ds-15_dOCR.fa > t && mv t Ds-15_dOCR.fa           
```
### fimo
```
fimo --o Ds-15_dOCR Pbr_TF_binding_motifs.meme Ds-15_dOCR.fa      
```
### Extracting and merging weighted networks
```
python tran_fimo_to_network\(dOCR\).py -f Ds-15_dOCR/fimo.tsv -p Ds-15_peak_gene_contact.id -o Ds-15_dOCR.row.network
```
```python
import pandas as pd
import numpy as np 
a = pd.read_csv('Ds-15_dOCR.row.network') 
a 
b = a['weight'] 
a['weight'] = (np.log2(b)-np.log2(min(b))+1)/(np.log2(max(b))-np.log2(min(b))+1) 
a.to_csv('Ds-15_dOCR.network',index=False)              
```
## 8.Merge to get ATAC_PWM_network
```
python merge_two_network.py -n1 Ds-15_promoter_peak.network -n2 Ds-15_dOCR.network -o ATAC_PWM_network.csv
awk -F',' '$3>0.8' ATAC_PWM_network.csv > ATAC_PWM_network_0.8.csv
```
