# ATAC-PWM Network
## 1.Structural annotations of peaks
### Preparation: Import the package, install it if you don't have it.  
In addition, you need to use gff/gtf to build a TxDb. Because the official website does not have TxDb for every species (currently 43 in total), we build it manually.
```R
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("clusterProfiler")

library(clusterProfiler)
library(ChIPseeker)
library(GenomicFeatures)
txdb <- makeTxDbFromGFF("DS.gff",
                        format="gff")
keytypes(txdb)
keys(txdb)
```
### Structural annotations for a single file
```R
#Read a single summits file
peaks <- readPeakFile("Pbr-4.7.idr.0.05.narrowPeak.sorted")
#Check the distribution of all peaks on each chromosome
covplot(peaks, weightCol="FE")
#Structural annotations
#options(ChIPseeker.downstreamDistance = 500)
#Set the length of the gene downstream
peakAnno <- annotatePeak(peaks,
                         TxDb=txdb,
                         level = "gene",
                         tssRegion=c(-2500, 2500))
#After annotation, visualize it and choose from a variety of pictures
plotAnnoBar(peakAnno)
plotDistToTSS(peakAnno)
vennpie(peakAnno)
plotAnnoPie(peakAnno)
#install.packages("ggupset")
library(ggupset)
upsetplot(peakAnno)
#install.packages("ggimage")
library(ggimage)
upsetplot(peakAnno, vennpie=TRUE)
#Finally, convert our annotation results into a data frame for easy viewing
df <- as.data.frame(peakAnno)
write.csv(df,file = '7.6_peakAnno.csv', quote = F, row.names = F)
#Extract the annotated genes (column 14) for subsequent functional analysis
gene <- df[,14]
```
### Structural annotations for multiple files
```R
#You can also read multiple summits files at a time, store them in list, and then annotate them using lapply
files = list(peak1 = "Pbr-4.7.idr.0.05.narrowPeak.sorted", peak2 = ("Pbr-7.6.idr.0.05.narrowPeak.sorted"))
peakAnnoList <- lapply(files, 
                       annotatePeak,
                       TxDb=txdb,
                       level = "gene",
                       tssRegion=c(-2500, 2500))
plotAnnoBar(peakAnnoList)
plotDistToTSS(peakAnnoList)
```
