# cPPI Network
## 1.data collection
Arabidopsis protein interaction data download:  
'AI_interactions.csv', 11374 protein interactions, from the article 'Evidence for Network Evolution in an Arabidopsis Interactome Map'  
'BIOGRID-ORGANISM-Arabidopsis_thaliana_Columbia-4.4.227.tab3.xls', 82616 (74126 after deduplication), from the BIOGRID database  
'TairProteinInteraction1.0.xls', 2655 (1715 after deduplication), from TAIR database, data before 20090527, all from literature, high reliability, version v1.0  
'TairProteinInteraction2.0.xls', 72266, from the TAIR database, a set of about 70,000 predicted protein-protein interactions, and about 3,000 experimentally determined (merging datasets from TAIR, IntAct-EBI, and BIND/BOND), v2.0 version  
## 2.Arrange and merge the four sets of data, using two columns of pure ID separated by tab
```
#Merge, remove duplicates
cat AI_interactions.txt BIOGRID_interactions.txt TairPI1.0.txt TairPI2.0.txt|sort|uniq|grep -v '-'>merge_interactions.txt
sort -u <(awk '{if ($1 < $2) print $1, $2; else print $2, $1}' merge_interactions.txt)|awk '{print $1"\t"$2}' >t && mv t merge_interactions.txt
wc -l merge_interactions.txt
142483 merge_interactions.txt
```
## 3.Identification of orthologous gene pairs
### orthofinder
Identification of Orthogroups in Arabidopsis and Pyrus using orthofinder.
### blast
```
diamond blastp -d Pbr.dmnd -q ../1.Ath_genome/Ath.pep -o Pbr-v1-Ath.blast -p 20 --sensitive --max-target-seqs 5 --evalue 1e-10 --quiet &
```
### Use Orthogroup and blast results to filter Best hits
In the orthogroup, the genes with similarity greater than 50, score greater than 100, and the highest score among the orthogroup genes were selected as best hits for conversion between Arabidopsis and pear genes.
```
python identify_ortholog.py --orthogroups Ath__v__Pbr.tsv --blast Pbr-v1-Ath.blast --output Ath_Pbr_hit --similarity 50
```
### Convert Arabidopsis IDs to Pyrus IDs
```
python tran_Ath_pep2Pbr_pep.py Ath_Pbr_hit merge_interactions.txt|sort|uniq>Pbr_interactions.csv
wc -l Pbr_interactions.csv
64003 Pbr_interactions.csv
#Pbr_interactions_pro.txt is the data used for database display
python tran_Ath_pep2Pbr_pep_pro.py Ath_Pbr_hit merge_interactions.txt|sort|uniq >Pbr_interactions_pro.txt
wc -l Pbr_interactions_pro.txt
73645 Pbr_interactions_pro.txt
```
## 4.Add source information
```
sort -u <(awk '{if ($1 < $2) print $1, $2; else print $2, $1}' AI_interactions.txt)|awk '{print $1"\t"$2}' >t && mv t AI_interactions.txt
sort -u <(awk '{if ($1 < $2) print $1, $2; else print $2, $1}' BIOGRID_interactions.txt)|awk '{print $1"\t"$2}' >t && mv t BIOGRID_interactions.txt
sort -u <(awk '{if ($1 < $2) print $1, $2; else print $2, $1}' TairPI1.0.txt)|awk '{print $1"\t"$2}' >t && mv t TairPI1.0.txt
sort -u <(awk '{if ($1 < $2) print $1, $2; else print $2, $1}' TairPI2.0.txt)|awk '{print $1"\t"$2}' >t && mv t TairPI2.0.txt
#Add source labels to protein interaction data
python add_PPI_label.py -A AI_interactions.txt -B BIOGRID_interactions.txt -T1 TairPI1.0.txt -T2 TairPI2.0.txt -P Pbr_interactions_pro.txt >Pbr_interactions_pro_label.txt
cut -f 1,3,6,8 ../1.Ath_data/TairProteinInteraction1.0.xls|sed '1d' >TairProteinInteraction1.0.txt
cut -f 6,7,10,11,12,15 BIOGRID-ORGANISM-Arabidopsis_thaliana_Columbia-4.4.227.tab3.xls|sed '1d' >BIOGRID.txt
cut -f 1,2,3,4 TairProteinInteraction2.0.xls|sed '1d' >TairProteinInteraction2.0.txt
# Sort and remove duplicates
awk 'BEGIN{FS="\t";OFS="\t"}{if ($1 < $2) print $1, $2,$3,$4,$5,$6; else print $2, $1,$3,$4,$5,$6}' BIOGRID.txt | sort -u |awk '$1!="-"' >t && mv t BIOGRID.txt
awk 'BEGIN{FS="\t";OFS="\t"}{if ($1 < $2) print $1, $2,$3,$4; else print $2, $1,$3,$4}' TairProteinInteraction1.0.txt | sort -u |awk '$1!="-"'  >t && mv t TairProteinInteraction1.0.txt
awk 'BEGIN{FS="\t";OFS="\t"}{if ($1 < $2) print $1, $2,$3,$4; else print $2, $1,$3,$4}' TairProteinInteraction1.0.txt | sort -u |awk '$1!="-"'  >t && mv t TairProteinInteraction1.0.txt
#Add more detailed comments
Generate a 9-column file: Pbr ID1\tAth ID1\tDescription1\tPbr ID2\tAth ID2\tDescription2\tDatabase Source\tPublication\tExperiment Source
```
## 5.annotate
```
python add_PPI_annotation.py -B ../1.Ath_data/BIOGRID.txt -T1 ../1.Ath_data/TairProteinInteraction1.0.txt -T2 ../1.Ath_data/TairProteinInteraction2.0.txt -P Pbr_interactions_pro_label.txt >Pbr_interactions_pro_annotation.txt
echo 'Pbr_ID1\tAth_ID1\tDescription1\tPbr_ID2\tAth_ID2\tDescription2\tDatabase_Source\tPublication\tExperiment_Source' >head
cat head Pbr_interactions_pro_annotation.txt >t && mv t Pbr_interactions_pro_annotation.txt
#Merge with Pcorr for Interact_Finder module retrieval
python merge_for_interact_finder.py Pcorr/Pcorr.annotations cPPI/Pbr_interactions_pro_annotation.txt >interact_finder.annotations
cat head interact_finder.annotations > t && mv t interact_finder.annotations
```
