# Nocoding Network
## 1.TPM
```
cp ../11.lncRNA/08.exp/linc.TPM .
cp ../12.circRNA/CIRI/07.quant/gtf/circRNA.CPM .
cp ../14.total_RNA/03.exp/TotalRNA.TPM .
cp ../13.sRNA/miRDP2/7.gff/miRNA.CPM .
head -1 TotalRNA.TPM >head
sed -i '1d' TotalRNA.TPM circRNA.CPM linc.TPM miRNA.CPM
cat head TotalRNA.TPM linc.TPM circRNA.CPM miRNA.CPM >all.TPM
```
## 2.WGCNA
```
mamba activate WGCNA
mamba install r-wgcna -c bioconda
# WGCNA version 1.71
nohup Rscript WGCNA.r 01.TPM/all.TPM SoftThreshold.pdf WGCNA.edge &
awk '$3>0.6' WGCNA.edge|cut -f 1,2,3 >WGCNA_0.6.edge
awk '{print $1"\t"$2"\t"$3"\tWGCNA"}' WGCNA_0.6.edge >t && mv t WGCNA_0.6.edge
```
## 3.TargetFinder（miRNA）
```
git clone https://gitee.com/lhx97/TargetFinder.git
#Install dependency fasta35
wget https://fasta.bioch.virginia.edu/wrpearson/fasta/fasta33-35/fasta-35.4.12.tar.gz
tar xzf fasta-35.4.12.tar.gz
cd fasta-35.4.12/src
make -f ../make/Makefile.linux_sse2 all
cd 13.sRNA/miRDP2/8.TargetFinder/result/
#Batch run TargetFinder
python ../run_TargetFinder.py ../miRNA.fa ../mRNA-v1.fa
#Merge results
for i in *;do cat $i >>../miRNA_predicted_targets.txt;done
cd 13.sRNA/miRDP2/8.TargetFinder
grep -v 'No results for' miRNA_predicted_targets.txt|cut -f 1,2 >t && mv t miRNA_predicted_targets.txt
sort miRNA_predicted_targets.txt|uniq > t && mv t miRNA_predicted_targets.txt
```
## 4.CircRNA-mRNA
```
cut -f 9 circRNA.gff3|sed 's/ID=//g'|sed 's/;Circ_type=.*;Gene_id=/\t/g'|sed 's/;//g' >circRNA-mRNA
python tran_circRNA_mRNA_interact.py circRNA-mRNA >circRNA.network
```
## 5.merge
```
python merge_two_networks.py -n1 WGCNA_0.6.edge -n2 circRNA.network -o t1
python merge_two_networks.py -n1 t1 -n2 miRNA_predicted_targets.txt -o ncRNA_network
python tran_direction.py Ds-v1_TF ncRNA_network >t && mv t ncRNA_network
```
